data/planet/hannes/re-developing-tcp-from-the-grounds-up.md

```diff
@@ -0,0 +1,49 @@
+<h1>The Secondary Nameserver Issue</h1>
+<p>One of our secondary nameservers attempts to receive zones (via AXFR using TCP) from another nameserver that is currently not running. Thus it replies to each SYN packet a corresponding RST. Below I graphed the network utilisation (send data/packets is positive y-axis, receive part on the negative) over time (on the x-axis) on the left and memory usage (bytes on y-axis) over time (x-axis) on the right of our nameserver. You can observe that both increases over time, and roughly every 3 hours, the unikernel hits its configured memory limit (64 MB), crashes with *out of memory*, and is restarted. The graph below is using the `mirage-tcpip` stack.</p>
+<p><a href="https://hannes.robur.coop/static/img/a.ns.mtcp.png"><img src="https://hannes.robur.coop/static/img/a.ns.mtcp.png" width="750"/></a></p>
+<p>Now, after switching over to &micro;TCP, graphed below, there's much less network utilisation and the memory limit is only reached after 36 hours, which is a great result. Though, still it is not very satisfying that the unikernel leaks memory. On their left side, both graphs contain a few hours of `mirage-tcpip`, and shortly after 20:00 on Nov 23rd, &micro;TCP got deployed.</p>
+<p><a href="https://hannes.robur.coop/static/img/a.ns.mtcp-utcp.png"><img src="https://hannes.robur.coop/static/img/a.ns.mtcp-utcp.png" width="750"/></a></p>
+<p>Investigating the involved parts showed that a TCP connection that was never established has been registered at the MirageOS layer, but the pure core does not expose an event from the received RST that the connection has been cancelled. This means the MirageOS layer piles up all the connection attempts, and it doesn't inform the application that the connection couldn't be established. Once this was well understood, developing the <a href="https://github.com/robur-coop/utcp/commit/67fc49468e6b75b96a481ebe44dd11ce4bb76e6c">required code changes</a> was straightforward. The graph shows that the fix was deployed at 15:25. The memory usage is constant afterwards, but the network utilisation increased enormously.</p>
+<p><a href="https://hannes.robur.coop/static/img/a.ns.utcp-ev.png"><img src="https://hannes.robur.coop/static/img/a.ns.utcp-ev.png" width="750"/></a></p>
+<p>Now, the network utilisation is unwanted. This was hidden by the application waiting forever that the TCP connection getting established. Our bug fix uncovered another issue--a tight loop:</p>
```

Comment:
```suggestion
<p>Now, the network utilisation is unwanted. This was hidden by the application waiting forever while the TCP connection getting established. Our bug fix uncovered another issue--a tight loop:</p>
```

This makes sense with "while" instead of "that," but it might not be accurate. If it's not, let's work on this sentence together because it's confusing as it's written.

Reply:
also this makes total sense, I applied your proposal.